File: fountain.py
import random
import math
from functools import reduce

def xor_bytes(b1, b2):
    return bytes(a ^ b for a, b in zip(b1, b2))

def get_robust_soliton_distribution(K, c=0.1, delta=0.5):
    """
    K: Number of source blocks
    c: Constant > 0
    delta: Prob of failure to decode
    """
    if K == 1:
        return [1.0]
    
    # rho[d]
    rho = [0.0] * (K + 1)
    rho[1] = 1.0 / K
    for d in range(2, K + 1):
        rho[d] = 1.0 / (d * (d - 1))
    
    # tau[d]
    R = c * math.log(K / delta) * math.sqrt(K)
    tau = [0.0] * (K + 1)
    if R > 0:
        limit = min(K, max(1, int(round(K / R))))
        for d in range(1, limit):
            tau[d] = R / (d * K)
        tau[limit] = R * math.log(R / delta) / K
    
    # mu[d]
    mu = [rho[d] + tau[d] for d in range(K + 1)]
    beta = sum(mu)
    return [m / beta for m in mu]

class LTEncoder:
    def __init__(self, data, block_size, c=0.1, delta=0.5):
        self.data = data
        self.block_size = block_size
        self.K = math.ceil(len(data) / block_size)
        self.blocks = [data[i*block_size:(i+1)*block_size].ljust(block_size, b'\x00') 
                       for i in range(self.K)]
        self.mu = get_robust_soliton_distribution(self.K, c, delta)
        self.degrees = list(range(1, self.K + 1))
        
    def generate_symbol(self, seed):
        prng = random.Random(seed)
        
        # Pick a degree d based on mu
        d = prng.choices(self.degrees, weights=self.mu[1:])[0]
        
        # Pick d distinct source indices
        indices = prng.sample(range(self.K), d)
        
        # XOR blocks
        symbol_data = reduce(xor_bytes, [self.blocks[i] for i in indices])
        
        return d, symbol_data

class LTDecoder:
    def __init__(self, K, block_size, c=0.1, delta=0.5):
        self.K = K
        self.block_size = block_size
        self.mu = get_robust_soliton_distribution(K, c, delta)
        self.degrees = list(range(1, K + 1))
        
        self.blocks = [None] * K
        self.num_recovered = 0
        self.symbols = [] # List of [indices_set, data]
        self.recovered_queue = [] # Queue of (index, data) to propagate
        
    def is_done(self):
        return self.num_recovered == self.K
    
    def add_symbol(self, seed, symbol_data):
        if self.is_done():
            return
        
        prng = random.Random(seed)
        d = prng.choices(self.degrees, weights=self.mu[1:])[0]
        indices = set(prng.sample(range(self.K), d))
        
        # Reduce symbol with already recovered blocks
        for i in range(self.K):
            if self.blocks[i] is not None and i in indices:
                symbol_data = xor_bytes(symbol_data, self.blocks[i])
                indices.remove(i)
        
        if not indices:
            return
            
        if len(indices) == 1:
            idx = list(indices)[0]
            if self.blocks[idx] is None:
                self.recovered_queue.append((idx, symbol_data))
                self._propagate()
        else:
            # Check if we already have this symbol (set of indices)
            # This is a bit slow but helps
            for existing_indices, _ in self.symbols:
                if existing_indices == indices:
                    return
            self.symbols.append([indices, symbol_data])

    def _propagate(self):
        while self.recovered_queue:
            index, data = self.recovered_queue.pop(0)
            if self.blocks[index] is not None:
                continue
                
            self.blocks[index] = data
            self.num_recovered += 1
            
            if self.is_done():
                return

            new_symbols = []
            for sym in self.symbols:
                indices, sym_data = sym
                if index in indices:
                    indices.remove(index)
                    sym[1] = xor_bytes(sym_data, data)
                    sym_data = sym[1]
                    if len(indices) == 1:
                        idx = list(indices)[0]
                        if self.blocks[idx] is None:
                            self.recovered_queue.append((idx, sym_data))
                    elif len(indices) > 1:
                        new_symbols.append(sym)
                else:
                    new_symbols.append(sym)
            self.symbols = new_symbols
            # After each block recovery, we might have new degree-1 symbols in the queue

    def reconstruct(self):
        if not self.is_done():
            return None
        return b''.join(self.blocks)

File: ios/SmokeSignalCore/Sources/SmokeSignalCore/Fountain.swift
import Foundation

func robustSolitonDistribution(K: Int, c: Double = 0.1, delta: Double = 0.5) -> [Double] {
    precondition(K >= 0)
    if K == 0 {
        return [1.0]
    }
    if K == 1 {
        // Index 0 unused, degree 1 has full weight.
        return [0.0, 1.0]
    }

    var rho = Array(repeating: 0.0, count: K + 1)
    rho[1] = 1.0 / Double(K)
    if K >= 2 {
        for d in 2...K {
            rho[d] = 1.0 / (Double(d) * Double(d - 1))
        }
    }

    let R = c * log(Double(K) / delta) * sqrt(Double(K))
    var tau = Array(repeating: 0.0, count: K + 1)
    if R > 0 {
        let limit = min(K, max(1, Int((Double(K) / R).rounded())))
        if limit > 1 {
            for d in 1..<limit {
                tau[d] = R / (Double(d) * Double(K))
            }
        }
        tau[limit] = R * log(R / delta) / Double(K)
    }

    var mu = Array(repeating: 0.0, count: K + 1)
    for d in 0...K {
        mu[d] = rho[d] + tau[d]
    }
    let beta = mu.reduce(0.0, +)
    return mu.map { $0 / beta }
}

func cdfFromDistribution(_ mu: [Double]) -> [Double] {
    // mu[0] is unused; degrees are 1..K
    if mu.count <= 2 {
        return [1.0]
    }
    let K = mu.count - 1
    var cdf: [Double] = []
    cdf.reserveCapacity(K)
    var acc = 0.0
    for d in 1...K {
        acc += mu[d]
        cdf.append(acc)
    }
    cdf[cdf.count - 1] = 1.0
    return cdf
}

func sampleDegree(cdf: [Double], prng: inout XorShift32) -> Int {
    let r = prng.nextDouble01()
    var low = 0
    var high = cdf.count - 1
    while low < high {
        let mid = (low + high) / 2
        if r < cdf[mid] {
            high = mid
        } else {
            low = mid + 1
        }
    }
    return low + 1
}

func sampleUniqueIndices(count: Int, upperBound: Int, prng: inout XorShift32) -> [Int] {
    precondition(count >= 0)
    precondition(upperBound >= 0)
    if count == 0 || upperBound == 0 {
        return []
    }
    if count >= upperBound {
        return Array(0..<upperBound)
    }
    var chosen = Set<Int>()
    chosen.reserveCapacity(count)
    while chosen.count < count {
        chosen.insert(prng.nextInt(upperBound: upperBound))
    }
    return Array(chosen)
}

public final class FountainEncoder {
    public let blockSize: Int
    public let K: Int

    private let blocks: [Data]
    private let degreeCDF: [Double]

    public init(data: Data, blockSize: Int, c: Double = 0.1, delta: Double = 0.5) {
        precondition(blockSize > 0)
        self.blockSize = blockSize
        self.K = (data.count + blockSize - 1) / blockSize

        if K == 0 {
            self.blocks = []
        } else {
            var tmp: [Data] = []
            tmp.reserveCapacity(K)
            for i in 0..<K {
                let start = i * blockSize
                let end = min(start + blockSize, data.count)
                var chunk = data.subdata(in: start..<end)
                if chunk.count < blockSize {
                    chunk.append(contentsOf: repeatElement(0, count: blockSize - chunk.count))
                }
                tmp.append(chunk)
            }
            self.blocks = tmp
        }

        let mu = robustSolitonDistribution(K: K, c: c, delta: delta)
        self.degreeCDF = cdfFromDistribution(mu)
    }

    public func generateSymbol(seed: UInt32) -> (degree: Int, data: Data) {
        if K == 0 {
            return (0, Data())
        }
        var prng = XorShift32(seed: seed)
        let degree = sampleDegree(cdf: degreeCDF, prng: &prng)
        let indices = sampleUniqueIndices(count: degree, upperBound: K, prng: &prng)

        var bytes = Array(blocks[indices[0]])
        if indices.count > 1 {
            for idx in indices.dropFirst() {
                xorInPlace(&bytes, with: blocks[idx])
            }
        }
        return (degree, Data(bytes))
    }
}

public final class FountainDecoder {
    public let blockSize: Int
    public let K: Int
    public private(set) var numRecovered: Int = 0

    private var blocks: [Data?]
    private var symbols: [(indices: Set<Int>, data: Data)] = []
    private var recoveredQueue: [(index: Int, data: Data)] = []
    private var queueCursor: Int = 0
    private let degreeCDF: [Double]

    public init(K: Int, blockSize: Int, c: Double = 0.1, delta: Double = 0.5) {
        precondition(K >= 0)
        precondition(blockSize > 0)
        self.K = K
        self.blockSize = blockSize
        self.blocks = Array(repeating: nil, count: K)
        let mu = robustSolitonDistribution(K: K, c: c, delta: delta)
        self.degreeCDF = cdfFromDistribution(mu)

        if K == 0 {
            self.numRecovered = 0
        }
    }

    public func isDone() -> Bool {
        numRecovered == K
    }

    public func addSymbol(seed: UInt32, symbolData: Data) {
        if isDone() || K == 0 {
            return
        }
        guard symbolData.count == blockSize else {
            return
        }

        var prng = XorShift32(seed: seed)
        let degree = sampleDegree(cdf: degreeCDF, prng: &prng)
        let indicesArr = sampleUniqueIndices(count: degree, upperBound: K, prng: &prng)
        var indices = Set(indicesArr)

        var reduced = symbolData
        if !indices.isEmpty {
            var toRemove: [Int] = []
            toRemove.reserveCapacity(indices.count)
            for idx in indices {
                if let block = blocks[idx] {
                    reduced = xorData(reduced, block)
                    toRemove.append(idx)
                }
            }
            for idx in toRemove {
                indices.remove(idx)
            }
        }

        if indices.isEmpty {
            return
        }

        if indices.count == 1, let idx = indices.first {
            if blocks[idx] == nil {
                recoveredQueue.append((idx, reduced))
                propagate()
            }
            return
        }

        for existing in symbols {
            if existing.indices == indices {
                return
            }
        }
        symbols.append((indices, reduced))
    }

    private func propagate() {
        while queueCursor < recoveredQueue.count {
            let (index, data) = recoveredQueue[queueCursor]
            queueCursor += 1

            if blocks[index] != nil {
                continue
            }
            blocks[index] = data
            numRecovered += 1

            if isDone() {
                return
            }

            var newSymbols: [(indices: Set<Int>, data: Data)] = []
            newSymbols.reserveCapacity(symbols.count)

            for var sym in symbols {
                if sym.indices.contains(index) {
                    sym.indices.remove(index)
                    sym.data = xorData(sym.data, data)

                    if sym.indices.count == 1, let last = sym.indices.first {
                        if blocks[last] == nil {
                            recoveredQueue.append((last, sym.data))
                        }
                        continue
                    }
                    if sym.indices.isEmpty {
                        continue
                    }
                }
                newSymbols.append(sym)
            }
            symbols = newSymbols
        }

        // Compact queue storage after batch propagation.
        if queueCursor > 0 {
            recoveredQueue.removeFirst(queueCursor)
            queueCursor = 0
        }
    }

    public func reconstruct() -> Data? {
        if K == 0 {
            return Data()
        }
        guard isDone() else { return nil }
        var out = Data()
        out.reserveCapacity(K * blockSize)
        for block in blocks {
            guard let b = block else { return nil }
            out.append(b)
        }
        return out
    }
}

